pam/pam-module-transaction-dummy.go

```diff
@@ -0,0 +1,295 @@
+		return nil
+	}
+	if env == "" {
+		return pam.BadItem
+	}
+	m.Env[env] = value
+	return nil
+}
+
+// GetEnv is used to retrieve a PAM environment variable.
+func (m *PamModuleTransactionDummy) GetEnv(name string) string {
+	if m.Env == nil {
+		return ""
+	}
+
+	return m.Env[name]
+}
+
+// GetEnvList returns a copy of the PAM environment as a map.
+func (m *PamModuleTransactionDummy) GetEnvList() (map[string]string, error) {
+	if m.Env == nil {
+		m.Env = make(map[string]string)
+	}
+	return m.Env, nil
+}
+
+// GetUser is similar to GetItem(User), but it would start a conversation if
+// no user is currently set in PAM.
+func (m *PamModuleTransactionDummy) GetUser(prompt string) (string, error) {
+	if user, err := m.GetItem(pam.User); err != nil {
+		return "", err
+	} else if user != "" {
+		return user, nil
+	}
+
+	resp, err := m.StartStringConv(pam.PromptEchoOn, prompt)
+	if err != nil {
+		return "", err
+	}
+
+	return resp.Response(), nil
+}
+
+// SetData allows to save any value in the module data that is preserved
+// during the whole time the module is loaded.
+func (m *PamModuleTransactionDummy) SetData(key string, data any) error {
+	if m.Data == nil {
+		m.Data = make(map[string]any)
+	}
+
+	m.Data[key] = data
+	return nil
+}
+
+// GetData allows to get any value from the module data saved using SetData
+// that is preserved across the whole time the module is loaded.
+func (m *PamModuleTransactionDummy) GetData(key string) (any, error) {
+	if m.Data == nil {
+		return nil, pam.NoModuleData
+	}
+
+	data, found := m.Data[key]
+	if found {
+		return data, nil
+	}
+	return nil, pam.NoModuleData
+}
+
+// DummyStringResponse is a simple implementation of pam.StringConvResponse.
+type DummyStringResponse struct {
+	style   pam.Style
+	content string
+}
+
+// Style returns the conversation style of the DummyStringResponse.
+func (s DummyStringResponse) Style() pam.Style {
+	return s.style
+}
+
+// Response returns the string response of the DummyStringResponse.
+func (s DummyStringResponse) Response() string {
+	return s.content
+}
+
+// DummyBinaryResponse is an implementation of pam.BinaryConvResponse.
+type DummyBinaryResponse struct {
+	ptr pam.BinaryPointer
+}
+
+// Style returns the response style for the response, so always BinaryPrompt.
+func (b DummyBinaryResponse) Style() pam.Style {
+	return pam.BinaryPrompt
+}
+
+// Data returns the response native pointer, it's up to the protocol to parse
+// it accordingly.
+func (b DummyBinaryResponse) Data() pam.BinaryPointer {
+	return b.ptr
+}
+
+// Decode decodes the binary data using the provided decoder function.
+func (b DummyBinaryResponse) Decode(decoder pam.BinaryDecoder) (
+	[]byte, error) {
+	if decoder == nil {
+		return nil, errors.New("nil decoder provided")
+	}
+	return decoder(b.ptr)
+}
+
+// StartStringConv starts a text-based conversation using the provided style
+// and prompt.
+func (m *PamModuleTransactionDummy) StartStringConv(style pam.Style, prompt string) (
```

Comment:
Can we move this closer to the type declaration? The definitions and implementations for the Responses are between some of the methods of `PamModuleTransactionDummy`.